### PR TITLE
[SYCLomatic] Support migration of 3 NCCL APIs and 2 types

### PR DESCRIPTION
**Pull request by ziranzha**

### CODE REVIEW

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -2540,8 +2539,9 @@ void TypeInDeclRule::registerMatcher(MatchFinder &MF) {
                   "cooperative_groups::__v1::thread_block",
                   "libraryPropertyType_t", "libraryPropertyType",
                   "cudaDataType_t", "cudaDataType", "cublasComputeType_t",
-                  "cublasAtomicsMode_t", "CUmem_advise_enum", "CUmem_advise"),
-              matchesName("nccl.*")))))))
+                  "cublasAtomicsMode_t", "CUmem_advise_enum", "CUmem_advise",
+                  "ncclUniqueId", "ncclComm_t")
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp]: let's put all nccl migratino rules into a seperate files.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 2]
```diff
@@ -17558,6 +17510,26 @@ void ComplexAPIRule::runRule(
 
 REGISTER_RULE(ComplexAPIRule)
 
+void NcclAPIRule::registerMatcher(ast_matchers::MatchFinder &MF) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp]: same as above

File: clang/test/dpct/helper_files_ref/include/ccl_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
+//==---- ccl_utils.hpp----------------------------*- C++ -*----------------==//
+//
+// Copyright (C) Intel Corporation
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __DPCT_CCL_HPP__
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/ccl_utils.hpp]: should be __DPCT__CCL_UTILS_HPP__

File: clang/test/dpct/helper_files_ref/include/ccl_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+/// Get stored kvs with specified kvs address.
+inline std::shared_ptr<oneapi::ccl::kvs> &
+get_kvs(const oneapi::ccl::kvs::address_type &addr) {
+  static std::unordered_map<oneapi::ccl::kvs::address_type,
+                            std::shared_ptr<oneapi::ccl::kvs>>
+      kvs_map;
+  return kvs_map[addr];
+}
+// DPCT_LABEL_END
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/ccl_utils.hpp]: Please make sure line 29 is not show up in this final help file.

File: clang/test/dpct/helper_files_ref/include/ccl_utils.hpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+} // namespace detail
+
+/// Get concatenated library version as an integer.
+static inline int get_version() {
+  auto ver = oneapi::ccl::get_library_version();
+  return ver.major * 10000 + ver.minor * 100 + ver.update;
+}
+// DPCT_LABEL_END
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/ccl_utils.hpp]: same as above.

File: clang/test/dpct/helper_files_ref/include/ccl_utils.hpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  auto &ptr = detail::get_kvs(addr);
+  if (!ptr)
+    ptr = oneapi::ccl::create_kvs(addr);
+  return ptr;
+}
+
+} // namespace ccl
+} // namespace dpct
+
+#endif // __DPCT_CCL_HPP__
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/ccl_utils.hpp]: Macro guard for file need to be updated.

File: clang/lib/DPCT/ASTTraversal.h [hunk 7]
```diff
@@ -1864,6 +1863,7 @@ class TextureRule : public NamedMigrationRule<TextureRule> {
 template <typename T> class RuleRegister {
 public:
   RuleRegister(const char *ID, const std::string &Name) {
+    std::cout<<"register rule: "<<Name<<std::endl;
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.h]: do we need this std::cout?

Comment (ziranzha) [reply clang/lib/DPCT/ASTTraversal.h]: Fixed

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Review by YuriPlyakhin — APPROVED** [review]

**Pull request merged**